### DISCUSSION TITLE
Focus trap doesn't seem to be working in Modal in React Aria Components

### DISCUSSION THREAD

**Discussion by MarekBodingerBA** [adobe/react-spectrum#4892] [Q&A] [answered]
Hello,

I didn't want to create the issue, because there might be some easily catchable issue in the code I am not able to see. Basically, my modal doesn't trap the focus. I cannot really say whether it's an library issue or my code issue.

Modal component:
https://github.com/bratislava/konto.bratislava.sk/blob/56a7ea961480d2b19f69de19d1440a865b7efe8e/next/components/forms/simple-components/ModalV2.tsx#L19

Modal component usage:
https://github.com/bratislava/konto.bratislava.sk/blob/56a7ea961480d2b19f69de19d1440a865b7efe8e/next/components/forms/segments/SkipStepModal/SkipStepModal.tsx#L16
https://github.com/bratislava/konto.bratislava.sk/blob/56a7ea961480d2b19f69de19d1440a865b7efe8e/next/components/styleguide/showcases/ModalShowCase.tsx#L284

**Reproduction:**
https://city-account-next.staging.bratislava.sk/styleguide#modal -> `Open skip step modal`

Thank you for help.

**Comment by LFDanLu** [comment]
From a glance, nothing really jumps out to me from your implementation. Mind logging out what value [contain](https://github.com/adobe/react-spectrum/blob/1775ff2a331be32c3b27417acc2705ad11019b37/packages/%40react-aria/overlays/src/Overlay.tsx#L49) is returning after you open the Modal? I'm expecting it to be true after the Modal renders since it should be called by useDialog

**Reply by MarekBodingerBA** [reply]
![chrome_wl5pHAcxIW](https://github.com/adobe/react-spectrum/assets/104828482/331f28af-2b26-4ff9-84ec-5de4fb599ae0)
For some reason it is false.

**Reply by MarekBodingerBA** [reply]
https://github.com/adobe/react-spectrum/assets/104828482/01e2dc36-4256-40a7-9371-abb254798ed7

The `ctx` in `useOverlayFocusContain` is always `null`.

**Reply by MarekBodingerBA** [reply]
![image](https://github.com/adobe/react-spectrum/assets/104828482/387befed-8016-4000-be2b-9f8bffaa972b)
This seems to be the reason, the hook which tries to get the OverlayContext is above the context itself in the component tree.

**Reply by MarekBodingerBA** [reply]
Context get:

https://github.com/adobe/react-spectrum/blob/be501251a2012b2c1401e79940cbb39151c84d9d/packages/react-aria-components/src/Modal.tsx#L131 -> https://github.com/adobe/react-spectrum/blob/be501251a2012b2c1401e79940cbb39151c84d9d/packages/%40react-aria/overlays/src/useModalOverlay.ts#L57 -> https://github.com/adobe/react-spectrum/blob/be501251a2012b2c1401e79940cbb39151c84d9d/packages/%40react-aria/overlays/src/Overlay.tsx#L78

Context set:
https://github.com/adobe/react-spectrum/blob/be501251a2012b2c1401e79940cbb39151c84d9d/packages/react-aria-components/src/Modal.tsx#L151

**Reply by LFDanLu** [reply]
Odd that the context is always null, `useOverlayFocusContain` should be also called by the [Dialog](https://github.com/adobe/react-spectrum/blob/be501251a2012b2c1401e79940cbb39151c84d9d/packages/%40react-aria/dialog/src/useDialog.ts#L62) which should have access to the `OverlayContext.Provider` you show in the screenshot above since it is a child of the Modal...

**Reply by LFDanLu** [reply, marked as answer]
Ah I checked your yarn.lock, it could be due to multiple versions of [@react-aria/overlay](https://github.com/bratislava/konto.bratislava.sk/blob/56a7ea961480d2b19f69de19d1440a865b7efe8e/next/yarn.lock#L4544-L4561), resulting in multiple versions of the `OverlayContext`. This means the OverlayContext setup by the Modal's Overlay might be a different one that the Dialog within tries to access, hence the `null` value for `ctx` in `useOverlayFocusContain`. Could you confirm if this might be the case?

**Reply by MarekBodingerBA** [reply]
Oh, thank you! I deleted the unnecessary `@react-aria/overlays` package and it works. 💪 

I would like to use the thread to also ask a related question. Is this needed for modal in SSR?
```
const isSSR = useIsSSR()

if (isSSR) {
  return null
}
```
The error is missing `window`, isn't there an internal check for SSR?
```
ReferenceError: window is not defined
    at $5df64b3807dc15ee$var$getViewportSize (file:///C:/Projects/konto.bratislava.sk/next/node_modules/@react-aria/utils/dist/import.mjs:811:172)
    at file:///C:/Projects/konto.bratislava.sk/next/node_modules/@react-aria/utils/dist/import.mjs:790:52
    at useReducer (C:\Projects\konto.bratislava.sk\next\node_modules\react-dom\cjs\react-dom-server.browser.development.js:5079:57)
    at Object.useState (C:\Projects\konto.bratislava.sk\next\node_modules\react-dom\cjs\react-dom-server.browser.development.js:5017:10)
```

**Reply by MarekBodingerBA** [reply]
Also, is [this](https://github.com/bratislava/konto.bratislava.sk/blob/56a7ea961480d2b19f69de19d1440a865b7efe8e/next/components/forms/simple-components/ModalV2.tsx#L41) `onClose` needed? Shouldn't it propagate to the modal?

**Reply by LFDanLu** [reply]
> I would like to use the thread to also ask a related question. Is this needed for modal in SSR?

Oh interesting, our `react-aria` [Overlay](https://github.com/adobe/react-spectrum/blob/be501251a2012b2c1401e79940cbb39151c84d9d/packages/%40react-aria/overlays/src/Overlay.tsx#L47-L54) already has SSR handling built in but RAC Modal still has a [hook call](https://github.com/adobe/react-spectrum/blob/be501251a2012b2c1401e79940cbb39151c84d9d/packages/react-aria-components/src/Modal.tsx#L144) that relies on the `window` existing which is a problem for SSR. We should either change Modal so that hook call + the div it renders is in a child component of `ModalOverlayInner` so that Overlay wouldn't render it server side and/or make `useViewportSize` have a `useIsSSR` check within it as well. Mind filing an issue for that? Ideally, we would handle is `useIsSSR` check in as many places as necessary so you wouldn't have to do so. 

> Also, is [this](https://github.com/bratislava/konto.bratislava.sk/blob/56a7ea961480d2b19f69de19d1440a865b7efe8e/next/components/forms/simple-components/ModalV2.tsx#L41) onClose needed? Shouldn't it propagate to the modal?

From a glance, that `onClose` is needed since you have a button that uses the `close` from the Dialog's render props which comes from either the Dialog's `props.onClose` applied on the Dialog directly or through a [DialogContext](https://github.com/adobe/react-spectrum/blob/be501251a2012b2c1401e79940cbb39151c84d9d/packages/react-aria-components/src/Dialog.tsx#L61) that a DialogTrigger provides. Since you are directly controlling the Modal's open state without a DialogTrigger, what you are doing is correct. Perhaps we could have the Dialog access a Modal/Popover's context as a fallback though, will have to think about that.

**Reply by MarekBodingerBA** [reply]
Thank you, I've created the issue:
https://github.com/adobe/react-spectrum/issues/4921
